vllm/engine/llm_engine.py

```diff
@@ -135,70 +132,125 @@ def _init_workers(self, distributed_init_method: str):
             "Ray is required if parallel_config.world_size > 1.")
 
         self.workers: List[Worker] = []
-        worker = Worker(
+        distributed_init_method = f"tcp://{get_ip()}:{get_open_port()}"
+        self.driver_worker = Worker(
             self.model_config,
             self.parallel_config,
             self.scheduler_config,
-            0,
-            distributed_init_method,
-        )
-        self.workers.append(worker)
-        self._run_workers(
-            "init_model",
-            get_all_outputs=True,
-        )
-        self._run_workers(
-            "load_model",
-            get_all_outputs=True,
-            max_concurrent_workers=self.parallel_config.
-            max_parallel_loading_workers,
+            local_rank=0,
+            rank=0,
+            distributed_init_method=distributed_init_method,
+            is_driver_worker=True,
         )
+        self._run_workers("init_model")
+        self._run_workers("load_model")
 
     def _init_workers_ray(self, placement_group: "PlacementGroup",
                           **ray_remote_kwargs):
-        # Lazy import the Worker to avoid importing torch.cuda/xformers
-        # before CUDA_VISIBLE_DEVICES is set in the Worker
-        from vllm.worker.worker import Worker
+        if self.parallel_config.tensor_parallel_size == 1:
+            num_gpus = self.cache_config.gpu_memory_utilization
+        else:
+            num_gpus = 1
 
-        self.workers: List[Worker] = []
-        for bundle in placement_group.bundle_specs:
+        self.driver_dummy_worker: RayWorkerVllm = None
+        self.workers: List[RayWorkerVllm] = []
+
+        for bundle_id, bundle in enumerate(placement_group.bundle_specs):
             if not bundle.get("GPU", 0):
                 continue
-            if self.parallel_config.tensor_parallel_size == 1:
-                num_gpus = self.cache_config.gpu_memory_utilization
-            else:
-                num_gpus = 1
+
+            scheduling_strategy = PlacementGroupSchedulingStrategy(
+                placement_group=placement_group,
+                placement_group_capture_child_tasks=True,
+                placement_group_bundle_index=bundle_id,
+            )
+
+            if (bundle.get("node:__internal_head__", 0) > 0
```

Comment:
I don't think this is a good idea as it will prevent using vLLM on clusters where the head node is CPU only, or with Ray Serve. I would instead suggest the following:
1. start one dummy worker for each bundle in the placement group
2. get node IPs of all the dummy workers
3. see if one of the dummy workers matches the node IP of the current process
4. if not, throw exception, otherwise, this dummy worker will be the "real" dummy worker
5. tear down all dummy workers and start real workers + the "real" dummy worker

this approach should be correct under any circumstance and allow for multiple instances of vLLM to correctly follow Ray scheduling contracts.

Reply:
iiuc, you would like to avoid the requirement of using `node:__internal_head__`. Instead, we can just check the IP address of each worker, if a worker's IP matches driver's IP, then set that worker to be driver's dummy worker. If no worker's IP matches the driver's IP, we throw an exception.

Is my understanding correct? If so, I can change the code accordingly.

Reply:
Yep, that's exactly it!

Reply:
@Yard1 Fixed. Can you take a look again?